README.md
Framework7 Indexed List Plugin
=============================
Demo: http://boynet.github.io/boynet/


This plugin is for Indexed-List scroll like iOS have (and Android), currently works only with Contacts List
![''](http://i58.tinypic.com/2608tmo.jpg)


## Installation

Just grab plugin files from `dist/` folder or using bower:

```
bower install framework7-indexed-list-plugin
```

And link them to your app's right AFTER Framework7's scripts and styles:

```
<link rel="stylesheet" href="path/to/framework7.min.css">
<link rel="stylesheet" href="path/to/framework7.indexed-list.css">
...
<script src="path/to/framework7.min.js"></script>
<script src="path/to/framework7.indexed-list.js"></script>
```

## Usage
 put this html code inside .page
````
<ul class="list-index"></ul>
````

## Demo:
http://boynet.github.io/boynet/

OR:

Plugin comes with demo example to see how it works and looks. To make demo works you need: 

* install bower dependencies. Go to `demo/` folder and execute in terminal `bower install`


CHANGELOG.md
# Change Log

## Framework7-Indexed-List-plugin v0.1.0 - Updated on 17 Sep 2014
  * Initial release. 


alphascroll.js
Framework7.prototype.plugins.indexedlist = function (app, params) {
    var $ = window.Dom7;
    

    function initIndexedList(page) {
        var eventsTarget = $(page.container).find('.list-index');
        if (eventsTarget.length === 0) return;
        
        var pageContent = $(page.container).find('.page-content');
        buildLetters();
        var isTouched, isMoved;
        var letterToScroll;
        var elementHover;
        var fixedNavbar = pageContent.parents('.navbar-fixed').length > 0 || pageContent.parents('.navbar-through').length > 0;
        function handleTouchStart(e) {
            e.preventDefault();
            isTouched = true;

            var target = $(e.target);
            if (!target.is('li')) target = target.parents('li');
            if (target.length > 0) {
                scrollToLetter(target.eq(0).data('index-letter'));
            }
        }

        function handleTouchMove(e) {
            if (!isTouched) return;
            e.preventDefault();
            var target;
            if (e.type === 'mousemove') {
                target = $(e.target);
            }
            else {
                target = $(document.elementFromPoint(e.touches[0].pageX, e.touches[0].pageY));
            }
            if (!target.is('li')) target = target.parents('li');
                
            if (target.length === 0) return;
            if (target.length > 0 && !target.is('.list-index li')) return;

            scrollToLetter(target.eq(0).data('index-letter'));
        }

        function handleTouchEnd(e) {
            isTouched = isMoved = false;
        }
        
        $(page.container).on('click','.list-index li',function(e){
            var target = $(e.target);
            if (!target.is('li')) target = target.parents('li');
            if (target.length > 0) {
                scrollToLetter(target.eq(0).data('index-letter'));
            }
        });

        function buildLetters(){
            var _letters = [];
            var lettersHtml = '';
            pageContent.find('.list-group').each(function () {
                var _letterDiv = $(this).find('ul .list-group-title');
                var _letter = _letterDiv.html().trim().charAt(0).toUpperCase();
                _letterDiv.attr('data-index-letter', _letter);
                lettersHtml += '<li data-index-letter="' + _letter + '">' + _letter + '</li>';
                _letters.push(_letter);
            });
            eventsTarget.html(lettersHtml);
            return _letters;
        }

        function scrollToLetter(letter) {
            var scrollToEl = pageContent.find('.list-group ul li[data-index-letter="' + letter + '"]');
            if (!scrollToEl.length) return;
            var scrollTop = scrollToEl.offset().top + pageContent.scrollTop() - (fixedNavbar ? 44 : 0);
            pageContent.scrollTop(scrollTop);
        }

        eventsTarget.on(app.touchEvents.start, handleTouchStart);
        eventsTarget.on(app.touchEvents.move, handleTouchMove);
        eventsTarget.on(app.touchEvents.end, handleTouchEnd);
    }


    return {
        hooks: {
            pageInit: initIndexedList,
        }
    };
};


Indexed-List.js
Framework7.prototype.plugins.Indexed-List = function (app, params) {
    var $ = window.Dom7;
    var letterToScroll, page;
    var letterscrolled;
    var elementHover;
    var letters=[];
    var eventsTarget;
    var pageContent;

    function initAlphascroll(page) {
        eventsTarget = $(page.container);
        if (!eventsTarget.hasClass('.Indexed-List')) {
            eventsTarget = eventsTarget.find('.Indexed-List');
        }
        if (eventsTarget.length === 0) return;
        page = page;
        pageContent = $$(page.container).find('.page-content');
        letters = getLetters();

        function handleTouchStart(e) {
            if (app.device.os === 'android') e.preventDefault();
        }

        function handleTouchMove(e) {
            elementHover = $$(document.elementFromPoint(e.touches[0].pageX, e.touches[0].pageY));
            if (!$$(elementHover).is('li')) {
                return;
            }
            letterToScroll = elementHover.html();
            hoverLetter();
        }

        function handleTouchEnd(e) {
            letterscrolled = null;
        }
        
        $$(pageContent).on('click','.Indexed-List li',function(e){
            elementHover = $$(e.target);
            letterToScroll = $$(e.target).html();
            hoverLetter();
        });

        function getLetters(){
            var _letters = [];
           pageContent.find('.list-group').each(function () {
               _letterDiv = $$(this).find('ul .list-group-title');
               _letter = _letterDiv.html();
               _letterDiv.addClass('letter-'+_letter);
               eventsTarget.append('<li>' + _letter + '</li>');
              _letters.push(_letter);
           });
            return _letters;
        }

        function hoverLetter() {
            if (letterToScroll !== letterscrolled) {
                pageContent.find('.Indexed-List li.hover').removeClass('hover');
                offset = -43;
                scrollTo = pageContent.find('.list-group ul li.letter-' + letterToScroll)
                if (!scrollTo.length) return;
                offset = offset + pageContent.offset().top + scrollTo.offset().top;
                pageContent.scrollTop(offset);
                letterscrolled = letterToScroll;
            }

        }

        eventsTarget.on(app.touchEvents.start, handleTouchStart);
        eventsTarget.on(app.touchEvents.move, handleTouchMove);
        eventsTarget.on(app.touchEvents.end, handleTouchEnd);
    }


    return {
        hooks: {
            pageInit:initAlphascroll,
            pageAfterAnimation: initAlphascroll
        }
    };
};


framework7.indexed-list.js
/*
 * Framework7 Indexed List 0.1.0
 * Framework7 plugin to add Alphabet indexed list
 *
 * https://github.com/boynet/Framework7-Indexed-List-plugin
 *
 * Copyright 2015, boynet
 *
 * Licensed under MIT
 *
 * Released on: 17 Sep, 2014
*/
Framework7.prototype.plugins.indexedlist = function (app, params) {
    var $ = window.Dom7;
    

    function initIndexedList(page) {
        var eventsTarget = $(page.container).find('.list-index');
        if (eventsTarget.length === 0) return;
        
        var pageContent = $(page.container).find('.page-content');
        buildLetters();
        var isTouched, isMoved;
        var letterToScroll;
        var elementHover;
        var fixedNavbar = pageContent.parents('.navbar-fixed').length > 0 || pageContent.parents('.navbar-through').length > 0;
        var searchBar = $(page.container).find('.searchbar').length > 0;
        
        if (searchBar){
            console.log(eventsTarget);
            eventsTarget.css('margin-top','44px');
        }
        
        function handleTouchStart(e) {
            e.preventDefault();
            isTouched = true;

            var target = $(e.target);
            if (!target.is('li')) target = target.parents('li');
            if (target.length > 0) {
                scrollToLetter(target.eq(0).data('index-letter'));
            }
        }

        function handleTouchMove(e) {
            if (!isTouched) return;
            e.preventDefault();
            var target;
            if (e.type === 'mousemove') {
                target = $(e.target);
            }
            else {
                target = $(document.elementFromPoint(e.touches[0].pageX, e.touches[0].pageY));
            }
            if (!target.is('li')) target = target.parents('li');
                
            if (target.length === 0) return;
            if (target.length > 0 && !target.is('.list-index li')) return;

            scrollToLetter(target.eq(0).data('index-letter'));
        }

        function handleTouchEnd(e) {
            isTouched = isMoved = false;
        }
        
        $(page.container).on('click','.list-index li',function(e){
            var target = $(e.target);
            if (!target.is('li')) target = target.parents('li');
            if (target.length > 0) {
                scrollToLetter(target.eq(0).data('index-letter'));
            }
        });

        function buildLetters(){
            var _letters = [];
            var lettersHtml = '';
            pageContent.find('.list-group').each(function () {
                var _letterDiv = $(this).find('ul .list-group-title');
                var _letter = _letterDiv.html().trim().charAt(0).toUpperCase();
                _letterDiv.attr('data-index-letter', _letter);
                lettersHtml += '<li data-index-letter="' + _letter + '">' + _letter + '</li>';
                _letters.push(_letter);
            });
            eventsTarget.html(lettersHtml);
            return _letters;
        }

        function scrollToLetter(letter) {
            var scrollToEl = pageContent.find('.list-group ul li[data-index-letter="' + letter + '"]');
            if (!scrollToEl.length) return;
            var scrollTop = scrollToEl.offset().top + pageContent.scrollTop() - (fixedNavbar ? 44 : 0) - (searchBar ? 44 : 0);
            pageContent.scrollTop(scrollTop);
        }

        eventsTarget.on(app.touchEvents.start, handleTouchStart);
        eventsTarget.on(app.touchEvents.move, handleTouchMove);
        eventsTarget.on(app.touchEvents.end, handleTouchEnd);
    }


    return {
        hooks: {
            pageInit: initIndexedList,
        }
    };
};
